File: frontend/utils/charts.py
"""
utils/charts.py
================
Plotly chart factory functions for all dashboard pages.
All charts use a consistent dark theme matching the CSS.
"""
from __future__ import annotations

import pandas as pd
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots
from typing import Optional

# ── Dark chart theme base layout ───────────────────────────────────────────
DARK_LAYOUT = dict(
    paper_bgcolor="rgba(15,20,25,0)",
    plot_bgcolor="rgba(26,31,46,0.6)",
    font=dict(family="Inter, sans-serif", color="#F5F5F5"),
    xaxis=dict(
        gridcolor="rgba(42,47,62,0.6)",
        zerolinecolor="rgba(42,47,62,0.8)",
        showgrid=True,
    ),
    yaxis=dict(
        gridcolor="rgba(42,47,62,0.6)",
        zerolinecolor="rgba(42,47,62,0.8)",
        showgrid=True,
    ),
    legend=dict(
        bgcolor="rgba(26,31,46,0.8)",
        bordercolor="rgba(42,47,62,0.8)",
        borderwidth=1,
    ),
    margin=dict(l=10, r=10, t=40, b=10),
    hovermode="x unified",
)

MODEL_COLORS = {
    "prophet":  "#1E88E5",
    "arima":    "#FFD700",
    "sarima":   "#00FF88",
    "lightgbm": "#FF6B35",
    "lstm":     "#B388FF",
}


def hex_to_rgba(hex_str: str, alpha: float = 0.1) -> str:
    h = hex_str.lstrip('#')
    r, g, b = tuple(int(h[i:i+2], 16) for i in (0, 2, 4))
    return f"rgba({r},{g},{b},{alpha})"


def _apply_dark(fig: go.Figure, title: str = "") -> go.Figure:
    layout = dict(**DARK_LAYOUT)
    if title:
        layout["title"] = dict(text=title, font=dict(size=15, color="#F5F5F5"), x=0)
    fig.update_layout(**layout)
    return fig


def forecast_chart(
    forecast_df: pd.DataFrame,
    historical_df: Optional[pd.DataFrame],
    category: str,
    model: str,
    show_history: bool = True,
) -> go.Figure:
    """Main forecast line chart with optional historical overlay."""
    fig = go.Figure()

    if show_history and historical_df is not None and not historical_df.empty:
        fig.add_trace(go.Scatter(
            x=historical_df["date"],
            y=historical_df["actual_sales"],
            name="Historical",
            line=dict(color="#1E88E5", width=2),
            opacity=0.8,
        ))

    color = MODEL_COLORS.get(model.lower(), "#00D9FF")
    fig.add_trace(go.Scatter(
        x=forecast_df["date"],
        y=forecast_df["prediction"],
        name=f"{model.upper()} Forecast",
        line=dict(color=color, width=2.5),
        fill="tozeroy",
        fillcolor=hex_to_rgba(color, 0.1),
    ))

    # Vertical separator line
    if show_history and historical_df is not None and not historical_df.empty:
        sep_date = historical_df["date"].iloc[-1] if len(historical_df) > 0 else None
        if sep_date:
            fig.add_vline(
                x=str(sep_date),
                line_dash="dash",
                line_color="rgba(176,181,192,0.5)",
                annotation_text="Forecast →",
                annotation_font_color="#B0B5C0",
            )

    return _apply_dark(fig, f"{category} — {model.upper()} Forecast")


def multi_model_forecast_chart(
    forecasts: dict[str, pd.DataFrame],
    historical_df: Optional[pd.DataFrame],
    category: str,
) -> go.Figure:
    """Overlay multiple model forecasts on one chart."""
    fig = go.Figure()

    if historical_df is not None and not historical_df.empty:
        fig.add_trace(go.Scatter(
            x=historical_df["date"],
            y=historical_df["actual_sales"],
            name="Actual",
            line=dict(color="#1E88E5", width=2, dash="dot"),
            opacity=0.7,
        ))

    for model, df in forecasts.items():
        if df is None or df.empty:
            continue
        color = MODEL_COLORS.get(model.lower(), "#B0B5C0")
        fig.add_trace(go.Scatter(
            x=df["date"],
            y=df["prediction"],
            name=model.upper(),
            line=dict(color=color, width=2),
        ))

    return _apply_dark(fig, f"{category} — All Models Comparison")


def anomaly_chart(
    results: list[dict],
    category: str,
    model: str,
) -> go.Figure:
    """Scatter+line chart highlighting anomalies by severity."""
    if not results:
        fig = go.Figure()
        return _apply_dark(fig, "No overlapping data for anomaly detection")

    df = pd.DataFrame(results)

    fig = go.Figure()

    # Actual sales line
    fig.add_trace(go.Scatter(
        x=df["date"], y=df["actual_sales"],
        name="Actual Sales",
        line=dict(color="#1E88E5", width=2),
    ))

    # Forecast line
    fig.add_trace(go.Scatter(
        x=df["date"], y=df["forecast_sales"],
        name="Forecast",
        line=dict(color="#FF6B35", width=2, dash="dot"),
    ))

    # Anomaly markers by severity
    sev_map = {
        "low":    ("#00FF88", "circle",         8),
        "medium": ("#FFD700", "circle",        10),
        "high":   ("#FF4444", "circle-open",   14),
    }
    for sev, (color, symbol, size) in sev_map.items():
        mask = df["severity"] == sev
        if mask.any():
            sub = df[mask]
            fig.add_trace(go.Scatter(
                x=sub["date"], y=sub["actual_sales"],
                mode="markers",
                name=f"{sev.capitalize()} Deviation",
                marker=dict(color=color, size=size, symbol=symbol,
                            line=dict(color=color, width=2)),
                hovertemplate="<b>%{x}</b><br>Actual: %{y:.2f}<extra></extra>",
            ))

    return _apply_dark(fig, f"{category} — {model.upper()} Anomaly Detection")


def what_if_chart(what_if_data: dict) -> go.Figure:
    """Dual-line chart comparing baseline vs adjusted forecast."""
    results = what_if_data.get("results", [])
    if not results:
        return _apply_dark(go.Figure(), "No data available")

    df = pd.DataFrame(results)

    fig = go.Figure()

    # Baseline
    fig.add_trace(go.Scatter(
        x=df["date"], y=df["baseline_sales"],
        name="Baseline Forecast",
        line=dict(color="#1E88E5", width=2.5),
    ))

    # Adjusted
    fig.add_trace(go.Scatter(
        x=df["date"], y=df["adjusted_sales"],
        name="Adjusted Forecast",
        line=dict(color="#00FF88", width=2.5),
        fill="tonexty",
        fillcolor="rgba(0,255,136,0.06)",
    ))

    # Disruption zone shading & zero markers
    disruption_days = what_if_data.get("disruption_days", 0)
    if disruption_days > 0:
        disrupted = [r for r in results if r["adjusted_sales"] == 0.0]
        if disrupted:
            x_start = disrupted[0]["date"]
            x_end = disrupted[-1]["date"]
            fig.add_vrect(
                x0=x_start, x1=x_end,
                fillcolor="rgba(255,68,68,0.18)",
                layer="below",
                line_width=1,
                line_color="rgba(255,68,68,0.6)",
                annotation_text=f"🚨 Supply Disruption ({disruption_days}d)",
                annotation_position="top left",
                annotation_font=dict(color="#FF4444", size=12, family="Inter, sans-serif"),
            )
            # Add zero-sales markers for disrupted points
            dis_df = pd.DataFrame(disrupted)
            fig.add_trace(go.Scatter(
                x=dis_df["date"], y=dis_df["adjusted_sales"],
                mode="markers",
                name="Stockout (0 units)",
                marker=dict(color="#FF4444", size=8, symbol="x"),
                hovertemplate="<b>%{x}</b><br>Stockout: 0 units<extra></extra>",
            ))

    return _apply_dark(fig, "What-If Scenario — Baseline vs Adjusted")


def mape_heatmap(metrics: dict) -> go.Figure:
    """Heatmap of MAE across categories × models."""
    categories = ["M01AB", "M01AE", "N02BA", "N02BE", "N05B", "N05C", "R03", "R06"]
    models = ["prophet", "arima", "sarima", "lightgbm", "lstm"]

    z = []
    for cat in categories:
        row = []
        cat_data = metrics.get(cat, {})
        for m in models:
            m_data = cat_data.get(m, {})
            v = m_data.get("MAE")
            row.append(float(v) if v is not None else None)
        z.append(row)

    fig = go.Figure(go.Heatmap(
        z=z,
        x=[m.upper() for m in models],
        y=categories,
        colorscale=[
            [0.0,  "#00FF88"],
            [0.35, "#FFD700"],
            [0.7,  "#FF6B35"],
            [1.0,  "#FF4444"],
        ],
        text=[[f"{v:.2f}" if v is not None else "N/A" for v in row] for row in z],
        texttemplate="%{text}",
        hovertemplate="<b>%{y}</b> · <b>%{x}</b><br>MAE: %{z:.2f}<extra></extra>",
        showscale=True,
        colorbar=dict(
            title=dict(text="MAE (units)", font=dict(color="#F5F5F5")),
            tickfont=dict(color="#F5F5F5"),
        ),
    ))
    return _apply_dark(fig, "MAE Heatmap — All Categories × Models")


def model_bar_chart(metrics: dict) -> go.Figure:
    """Average MAE per model across all categories."""
    categories = ["M01AB", "M01AE", "N02BA", "N02BE", "N05B", "N05C", "R03", "R06"]
    models = ["prophet", "arima", "sarima", "lightgbm", "lstm"]

    avg_maes = []
    for m in models:
        vals = []
        for cat in categories:
            cat_data = metrics.get(cat, {})
            m_data = cat_data.get(m, {})
            v = m_data.get("MAE")
            if v is not None:
                vals.append(float(v))
        avg_maes.append(round(sum(vals) / len(vals), 2) if vals else 0)

    fig = go.Figure(go.Bar(
        x=[m.upper() for m in models],
        y=avg_maes,
        marker=dict(
            color=[MODEL_COLORS.get(m, "#B0B5C0") for m in models],
            line=dict(color="rgba(255,255,255,0.1)", width=1),
        ),
        text=[f"{v:.2f}" for v in avg_maes],
        textposition="outside",
        textfont=dict(color="#F5F5F5"),
        hovertemplate="<b>%{x}</b><br>Avg MAE: %{y:.2f}<extra></extra>",
    ))
    fig.update_layout(yaxis_title="Average MAE (units)")
    return _apply_dark(fig, "Average MAE by Model (All 8 Categories)")


def severity_donut(results: list[dict]) -> go.Figure:
    """Donut chart of anomaly severity distribution."""
    counts = {"normal": 0, "moderate": 0, "anomaly-medium": 0, "anomaly-high": 0}
    for r in results:
        s, sev = r.get("status", ""), r.get("severity", "")
        if s == "normal":
            counts["normal"] += 1
        elif s == "moderate":
            counts["moderate"] += 1
        elif s == "anomaly" and sev == "medium":
            counts["anomaly-medium"] += 1
        elif s == "anomaly" and sev == "high":
            counts["anomaly-high"] += 1

    labels = list(counts.keys())
    values = list(counts.values())
    colors = ["#00FF88", "#FFD700", "#FF6B35", "#FF4444"]

    fig = go.Figure(go.Pie(
        labels=labels, values=values,
        hole=0.55,
        marker=dict(colors=colors, line=dict(color="rgba(15,20,25,0.8)", width=2)),
        textinfo="percent+label",
        textfont=dict(color="#F5F5F5", size=11),
        hovertemplate="<b>%{label}</b><br>Count: %{value}<br>%{percent}<extra></extra>",
    ))
    fig.update_layout(showlegend=False)
    return _apply_dark(fig, "Severity Distribution")

File: frontend-react/src/lib/charts.ts
// Plotly chart factories. Same trace types as the original Streamlit charts
// (line+fill, heatmap, donut, vrect/vline shading) but restyled to the
// muted premium palette — quiet gridlines, no neon fills.

import type { Data, Layout, Shape, Annotations } from 'plotly.js'
import { MODEL_COLORS } from './constants'
import type { AnomalyResult, ForecastPoint, WhatIfPoint, MetricsResponse } from '@/types/api'
import { CATEGORIES } from './constants'

export const DARK_LAYOUT: Partial<Layout> = {
  paper_bgcolor: 'rgba(0,0,0,0)',
  plot_bgcolor: 'rgba(0,0,0,0)',
  font: { family: 'Switzer, sans-serif', color: '#F4F4F5' },
  xaxis: { gridcolor: 'rgba(255,255,255,0.06)', zerolinecolor: 'rgba(255,255,255,0.1)', showgrid: true },
  yaxis: { gridcolor: 'rgba(255,255,255,0.06)', zerolinecolor: 'rgba(255,255,255,0.1)', showgrid: true },
  legend: { bgcolor: 'rgba(0,0,0,0)', bordercolor: 'rgba(255,255,255,0.08)', borderwidth: 1 },
  margin: { l: 10, r: 10, t: 40, b: 10 },
  hovermode: 'x unified',
  hoverlabel: { bgcolor: '#1B1B20', bordercolor: 'rgba(255,255,255,0.1)', font: { color: '#F4F4F5' } },
}

export function hexToRgba(hex: string, alpha = 0.1): string {
  const h = hex.replace('#', '')
  const r = parseInt(h.slice(0, 2), 16)
  const g = parseInt(h.slice(2, 4), 16)
  const b = parseInt(h.slice(4, 6), 16)
  return `rgba(${r},${g},${b},${alpha})`
}

export function applyDark(layout: Partial<Layout> = {}, title?: string): Partial<Layout> {
  return {
    ...DARK_LAYOUT,
    ...layout,
    ...(title ? { title: { text: title, font: { size: 15, color: '#F4F4F5' }, x: 0 } } : {}),
  }
}

export function forecastChart(
  forecast: ForecastPoint[],
  historical: { date: string; actual_sales: number }[] | null,
  category: string,
  model: string,
  showHistory: boolean,
): { data: Data[]; layout: Partial<Layout> } {
  const data: Data[] = []

  if (showHistory && historical && historical.length > 0) {
    data.push({
      x: historical.map((h) => h.date),
      y: historical.map((h) => h.actual_sales),
      name: 'Historical',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#6EA8D8', width: 2 },
      opacity: 0.8,
    })
  }

  const color = MODEL_COLORS[model.toLowerCase()] ?? '#7C6FF0'

  // Confidence band (only present for prophet/arima/sarima)
  const hasBounds = forecast.some((p) => p.lower !== undefined && p.upper !== undefined)
  if (hasBounds) {
    data.push({
      x: forecast.map((p) => p.date),
      y: forecast.map((p) => p.upper ?? p.prediction),
      name: 'Upper bound',
      type: 'scatter',
      mode: 'lines',
      line: { width: 0 },
      showlegend: false,
      hoverinfo: 'skip',
    })
    data.push({
      x: forecast.map((p) => p.date),
      y: forecast.map((p) => p.lower ?? p.prediction),
      name: 'Confidence Interval',
      type: 'scatter',
      mode: 'lines',
      line: { width: 0 },
      fill: 'tonexty',
      fillcolor: hexToRgba(color, 0.15),
      hoverinfo: 'skip',
    })
  }

  data.push({
    x: forecast.map((p) => p.date),
    y: forecast.map((p) => p.prediction),
    name: `${model.toUpperCase()} Forecast`,
    type: 'scatter',
    mode: 'lines',
    line: { color, width: 2.5 },
    fill: 'tozeroy',
    fillcolor: hexToRgba(color, 0.1),
  })

  const shapes: Partial<Shape>[] = []
  const annotations: Partial<Annotations>[] = []
  if (showHistory && historical && historical.length > 0) {
    const sepDate = historical[historical.length - 1].date
    shapes.push({
      type: 'line',
      x0: sepDate, x1: sepDate, y0: 0, y1: 1, yref: 'paper',
      line: { dash: 'dash', color: 'rgba(176,181,192,0.5)' },
    })
    annotations.push({ x: sepDate, y: 1, yref: 'paper', text: 'Forecast →', showarrow: false, font: { color: '#98989F' } })
  }

  return { data, layout: applyDark({ shapes, annotations }, `${category} — ${model.toUpperCase()} Forecast`) }
}

export function multiModelForecastChart(
  forecastsByModel: Record<string, ForecastPoint[]>,
  historical: { date: string; actual_sales: number }[] | null,
  category: string,
): { data: Data[]; layout: Partial<Layout> } {
  const data: Data[] = []

  if (historical && historical.length > 0) {
    data.push({
      x: historical.map((h) => h.date),
      y: historical.map((h) => h.actual_sales),
      name: 'Actual',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#6EA8D8', width: 2, dash: 'dot' },
      opacity: 0.7,
    })
  }

  for (const [model, points] of Object.entries(forecastsByModel)) {
    if (!points || points.length === 0) continue
    const color = MODEL_COLORS[model.toLowerCase()] ?? '#98989F'
    data.push({
      x: points.map((p) => p.date),
      y: points.map((p) => p.prediction),
      name: model.toUpperCase(),
      type: 'scatter',
      mode: 'lines',
      line: { color, width: 2 },
    })
  }

  return { data, layout: applyDark({}, `${category} — All Models Comparison`) }
}

const SEVERITY_MARKERS: Record<string, { color: string; symbol: string; size: number }> = {
  low: { color: '#3ECF8E', symbol: 'circle', size: 8 },
  medium: { color: '#F0B429', symbol: 'circle', size: 10 },
  high: { color: '#F0596B', symbol: 'circle-open', size: 14 },
}

export function anomalyChart(results: AnomalyResult[], category: string, model: string): { data: Data[]; layout: Partial<Layout> } {
  if (!results || results.length === 0) {
    return { data: [], layout: applyDark({}, 'No overlapping data for anomaly detection') }
  }

  const data: Data[] = [
    {
      x: results.map((r) => r.date),
      y: results.map((r) => r.actual_sales),
      name: 'Actual Sales',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#6EA8D8', width: 2 },
    },
    {
      x: results.map((r) => r.date),
      y: results.map((r) => r.forecast_sales),
      name: 'Forecast',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#E8836C', width: 2, dash: 'dot' },
    },
  ]

  for (const [sev, marker] of Object.entries(SEVERITY_MARKERS)) {
    const sub = results.filter((r) => r.severity === sev)
    if (sub.length === 0) continue
    data.push({
      x: sub.map((r) => r.date),
      y: sub.map((r) => r.actual_sales),
      name: `${sev[0].toUpperCase()}${sev.slice(1)} Deviation`,
      type: 'scatter',
      mode: 'markers',
      marker: { color: marker.color, size: marker.size, symbol: marker.symbol, line: { color: marker.color, width: 2 } },
    })
  }

  return { data, layout: applyDark({}, `${category} — ${model.toUpperCase()} Anomaly Detection`) }
}

export function whatIfChart(results: WhatIfPoint[], disruptionDays: number): { data: Data[]; layout: Partial<Layout> } {
  if (!results || results.length === 0) {
    return { data: [], layout: applyDark({}, 'No data available') }
  }

  const data: Data[] = [
    {
      x: results.map((r) => r.date),
      y: results.map((r) => r.baseline_sales),
      name: 'Baseline Forecast',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#6EA8D8', width: 2.5 },
    },
    {
      x: results.map((r) => r.date),
      y: results.map((r) => r.adjusted_sales),
      name: 'Adjusted Forecast',
      type: 'scatter',
      mode: 'lines',
      line: { color: '#3ECF8E', width: 2.5 },
      fill: 'tonexty',
      fillcolor: 'rgba(62,207,142,0.06)',
    },
  ]

  const shapes: Partial<Shape>[] = []
  const annotations: Partial<Annotations>[] = []

  if (disruptionDays > 0) {
    const disrupted = results.filter((r) => r.adjusted_sales === 0)
    if (disrupted.length > 0) {
      const x0 = disrupted[0].date
      const x1 = disrupted[disrupted.length - 1].date
      shapes.push({
        type: 'rect', x0, x1, y0: 0, y1: 1, yref: 'paper',
        fillcolor: 'rgba(240,89,107,0.18)', line: { width: 1, color: 'rgba(240,89,107,0.6)' }, layer: 'below',
      })
      annotations.push({
        x: x0, y: 1, yref: 'paper', xanchor: 'left',
        text: `🚨 Supply Disruption (${disruptionDays}d)`, showarrow: false,
        font: { color: '#F0596B', size: 12, family: 'Switzer, sans-serif' },
      })
      data.push({
        x: disrupted.map((r) => r.date),
        y: disrupted.map((r) => r.adjusted_sales),
        name: 'Stockout (0 units)',
        type: 'scatter',
        mode: 'markers',
        marker: { color: '#F0596B', size: 8, symbol: 'x' },
      })
    }
  }

  return { data, layout: applyDark({ shapes, annotations }, 'What-If Scenario — Baseline vs Adjusted') }
}

export function maeHeatmap(metrics: MetricsResponse): { data: Data[]; layout: Partial<Layout> } {
  const models = ['prophet', 'arima', 'sarima', 'lightgbm', 'lstm']
  const z = CATEGORIES.map((cat) => models.map((m) => {
    const v = metrics[cat]?.[m as keyof (typeof metrics)[string]] as { MAE?: number } | undefined
    return v?.MAE ?? null
  }))

  const data: Data[] = [{
    z,
    x: models.map((m) => m.toUpperCase()),
    y: [...CATEGORIES],
    type: 'heatmap',
    colorscale: [
      [0.0, '#3ECF8E'],
      [0.35, '#F0B429'],
      [0.7, '#E8836C'],
      [1.0, '#F0596B'],
    ],
    text: z.map((row) => row.map((v) => (v !== null ? v.toFixed(2) : 'N/A'))) as unknown as string[],
    texttemplate: '%{text}',
    hovertemplate: '<b>%{y}</b> · <b>%{x}</b><br>MAE: %{z:.2f}<extra></extra>',
    showscale: true,
    colorbar: { title: { text: 'MAE (units)', font: { color: '#F4F4F5' } }, tickfont: { color: '#F4F4F5' } },
  }]

  return { data, layout: applyDark({ margin: { l: 70, r: 10, t: 40, b: 10 } }, 'MAE Heatmap — All Categories × Models') }
}

export function modelBarChart(metrics: MetricsResponse): { data: Data[]; layout: Partial<Layout> } {
  const models = ['prophet', 'arima', 'sarima', 'lightgbm', 'lstm']
  const avgMaes = models.map((m) => {
    const vals = CATEGORIES
      .map((cat) => (metrics[cat]?.[m as keyof (typeof metrics)[string]] as { MAE?: number } | undefined)?.MAE)
      .filter((v): v is number => v !== undefined)
    return vals.length ? Math.round((vals.reduce((a, b) => a + b, 0) / vals.length) * 100) / 100 : 0
  })

  const data: Data[] = [{
    x: models.map((m) => m.toUpperCase()),
    y: avgMaes,
    type: 'bar',
    marker: { color: models.map((m) => MODEL_COLORS[m] ?? '#98989F'), line: { color: 'rgba(255,255,255,0.1)', width: 1 } },
    text: avgMaes.map((v) => v.toFixed(2)),
    textposition: 'outside',
    textfont: { color: '#F4F4F5' },
    hovertemplate: '<b>%{x}</b><br>Avg MAE: %{y:.2f}<extra></extra>',
  }]

  return { data, layout: applyDark({ yaxis: { ...DARK_LAYOUT.yaxis, title: { text: 'Average MAE (units)' } } }, 'Average MAE by Model (All 8 Categories)') }
}

export function severityDonut(results: AnomalyResult[]): { data: Data[]; layout: Partial<Layout> } {
  const counts = { normal: 0, moderate: 0, 'anomaly-medium': 0, 'anomaly-high': 0 }
  for (const r of results) {
    if (r.status === 'normal') counts.normal++
    else if (r.status === 'moderate') counts.moderate++
    else if (r.status === 'anomaly' && r.severity === 'medium') counts['anomaly-medium']++
    else if (r.status === 'anomaly' && r.severity === 'high') counts['anomaly-high']++
  }

  const labels = Object.keys(counts)
  const values = Object.values(counts)
  const colors = ['#3ECF8E', '#F0B429', '#E8836C', '#F0596B']

  const data: Data[] = [{
    labels, values,
    type: 'pie',
    hole: 0.55,
    marker: { colors, line: { color: 'rgba(15,20,25,0.8)', width: 2 } },
    textinfo: 'label+percent',
    textfont: { color: '#F4F4F5', size: 11 },
    hovertemplate: '<b>%{label}</b><br>Count: %{value}<br>%{percent}<extra></extra>',
  }]

  return { data, layout: applyDark({ showlegend: false }, 'Severity Distribution') }
}
